### DISCUSSION TITLE
Auto-open nvim-tree after opening a file

### DISCUSSION THREAD

**Discussion by musjj** [nvim-tree/nvim-tree.lua#1517] [Q&A]
I want to have nvim-tree auto-open after reading a file for the first time in the session.

I currently have this, and it works *almost* perfectly:

```lua
vim.api.nvim_create_autocmd({"BufNewFile", "BufReadPost"}, {
  callback = function(args)
    if vim.fn.expand "%:p" ~= "" then
      vim.api.nvim_del_autocmd(args.id)
      vim.cmd "noautocmd NvimTreeOpen"
      vim.cmd "noautocmd wincmd p"
    end
  end,
})
```

But `noautocmd wincmd p` fails to work when you launch nvim with a file (`nvim ~/file`), so the cursor ended up being focused on nvim-tree instead of the file buffer. Strangely, it works just fine if you open nvim without a file (`nvim`), and *then* read a file (`:e ~/file`).
What is causing this to fail?

EDIT: Figured out a hacky solution. Wrap the `wincmd` command with `vim.schedule`:
```lua
vim.schedule(function()
  vim.cmd "wincmd p"
end)
```
Not sure why it works, but whatever.

**Comment by alex-courtis** [comment]
You should ensure that `open_on_setup`, `open_on_setup_file`, `open_on_tab` are false so that they don't interfere with your automation.

Try `:lua require('nvim-tree').toggle(false, true)` to open the tree without focus. See `:help nvim-tree-api`. That will remove the need for `wincmd p`.

Your automation doesn't have any provision for "first time in the session"; it will always run. Options:
* global bool to record whether the first open has occurred
* unregister the autocmd after running the first time

**Reply by musjj** [reply]
Yes, I have those options turned off (removing the autocmd above will not open `nvim-tree` when opening a file).
The problem is not `nvim-tree` not opening, but `wincmd p` not working. I wanted nvim to focus on the file buffer instead of the filetree. 
`wincmd p`  works, when I open a file *after* launching nvim without any file arguments. But if I pass a file directly to nvim, it will launch with the filetree focused.
> unregister the autocmd after running the first time

Yes, I have this line: `vim.api.nvim_del_autocmd(args.id)` in the autocmd above

**Reply by alex-courtis** [reply]
> The problem is not nvim-tree not opening, but wincmd p not working

Did you try toggle? With the second argument true the tree should not be focused.

> Yes, I have this line: `vim.api.nvim_del_autocmd(args.id)` in the autocmd above

Whoops; I did not read properly.

**Reply by musjj** [reply]
Ah, forget to mention that. Yes I tried the toggle function and the same issue still occurs. It works when you `nvim` -> `:e ~/file`, but not when you `nvim ~/file`.
I suspect that this might be some kind of race condition? Maybe there is a need to wait for some kind of event before doing `wincmd p` on startup?

**Reply by alex-courtis** [reply]
> I suspect that this might be some kind of race condition? Maybe there is a need to wait for some kind of event before doing `wincmd p` on startup?

There are definitely race conditions at play when it comes to Buf* events. We removed the auto_close option for that reason: https://github.com/kyazdani42/nvim-tree.lua/pull/1466#issuecomment-1200339877 #1005 

> It works when you nvim -> :e ~/file, but not when you nvim ~/file.

It could be that the "BufNewFile", "BufReadPost" events are not fired when you `nvim ~/file`

Here's some hacky logging I use to examine events: https://github.com/kyazdani42/nvim-tree.lua/pull/1466#issuecomment-1200339877

I'm seeing some complex workarounds using open_on_setup_file and enabling your autocommand if the tree wasn't opened, however that sounds fragile.

I'm afraid I don't have anything further I can suggest.

**Reply by musjj** [reply]
> It could be that the "BufNewFile", "BufReadPost" events are not fired when you nvim ~/file

Sorry for not being clear, but the event was fired (`nvim-tree` gets toggled correctly), but the `wincmd p` command does not seem to work properly (the filetree is focused instead of the file buffer). What I suspect that maybe you need to wait for some kind of initial window operations to complete before you call `wincmd`.
